yarn-project/acir-simulator/src/avm/opcodes/external_calls.test.ts

```diff
@@ -0,0 +1,126 @@
+import { HostStorage } from '../journal/host_storage.js';
+import { AvmJournal } from '../journal/journal.js';
+import { encodeToBytecode } from './encode_to_bytecode.js';
+import { Call } from './external_calls.js';
+import { Opcode } from './opcodes.js';
+
+describe('External Calls', () => {
+  let machineState: AvmMachineState;
+  let journal: AvmJournal;
+
+  let contractsDb: MockProxy<PublicContractsDB>;
+
+  beforeEach(() => {
+    machineState = new AvmMachineState([], initExecutionEnvironment());
+
+    contractsDb = mock<PublicContractsDB>();
+
+    const commitmentsDb = mock<CommitmentsDB>();
+    const publicStateDb = mock<PublicStateDB>();
+    const hostStorage = new HostStorage(publicStateDb, contractsDb, commitmentsDb);
+    journal = new AvmJournal(hostStorage);
+  });
+
+  describe('Call', () => {
+    // TODO(https://github.com/AztecProtocol/aztec-packages/issues/3992): gas not implemented
+    it('Should execute a call correctly', async () => {
+      const gasOffset = 0;
+      const gas = Fr.zero();
+
+      const addrOffset = 1;
+      const addr = new Fr(123456n);
+
+      const argsOffset = 2;
+      const args = [new Fr(1n), new Fr(2n), new Fr(3n)];
+      const argsSize = args.length;
+
+      const retOffset = 8;
+      const retSize = 2;
+
+      const successOffset = 7;
+
+      machineState.writeMemory(0, gas);
+      machineState.writeMemory(1, addr);
+      machineState.writeMemoryChunk(2, args);
+
+      const otherContextInstructions: [Opcode, any[]][] = [
+        // Place [1,2,3] into memory
+        [Opcode.CALLDATACOPY, [/* value */ 0, /* copySize*/ argsSize, /* destOffset */ 0]],
+        // Store 1 into slot 1
+        [Opcode.SSTORE, [/* slotOffset */ 0, /* dataOffset */ 0]],
+        // Return [1,2] from memory
+        [Opcode.RETURN, [/* retOffset */ 0, /* size */ 2]],
+      ];
+
+      const otherContextInstructionsBytecode = Buffer.concat(
+        otherContextInstructions.map(([opcode, args]) => encodeToBytecode(opcode, args)),
+      );
+      jest
+        .spyOn(journal.hostStorage.contractsDb, 'getBytecode')
+        .mockReturnValue(Promise.resolve(otherContextInstructionsBytecode));
+
+      const instruction = new Call(gasOffset, addrOffset, argsOffset, argsSize, retOffset, retSize, successOffset);
+      await instruction.execute(machineState, journal);
+
+      const successValue = machineState.readMemory(successOffset);
+      expect(successValue).toEqual(new Fr(1n));
+
+      const retValue = machineState.readMemoryChunk(retOffset, retSize);
+      expect(retValue).toEqual([new Fr(1n), new Fr(2n)]);
+
+      // Check that the storage call has been merged into the parent journal
+      const { storageWrites } = journal.flush();
+      expect(storageWrites.size).toEqual(1);
+      const nestedContractWrites = storageWrites.get(addr);
+      expect(nestedContractWrites).toBeDefined();
+      expect(nestedContractWrites!.get(args[0])).toEqual(args[0]);
```

Comment:
[optional] I cannot see the type of this. Is it possible to specify the whole object? E.g. sth like: `expect(a).ToBe({...})` instead of indexing.

Reply:
I did a little refactor on the types and changed this up